scheduler_perf: dynamic resource allocation test cases

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Scheduling pods which use dynamic resource allocation currently suffers from pods being moved into the backoff queue because of false positives (see https://github.com/kubernetes/kubernetes/pull/115796). The workloads defined here can be used to measure that impact and try out fixes for it.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/assign @kerthcet

pohly committed:
scheduler perf: add DynamicResourceAllocation test cases

The default scheduler configuration must be based on the v1 API where the
plugin is enabled by default. Then if (and only if) the
DynamicResourceAllocation feature gate for a test is set, the corresponding
API group also gets enabled.

The normal dynamic resource claim controller is started if needed to create
ResourceClaims from ResourceClaimTemplates.

Without the upcoming optimizations in the scheduler, scheduling with dynamic
resources is fairly slow. The new test cases take around 15 minutes wall clock
time on my desktop.

pohly commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -758,6 +758,19 @@ func runWorkload(ctx context.Context, b *testing.B, tc *testCase, w *workload) [
 	// All namespaces listed in numPodsScheduledPerNamespace will be cleaned up.
 	numPodsScheduledPerNamespace := make(map[string]int)
+
+	maybeCreateNamespace := func(namespace string) {
```
This is only used once now, but will be needed more often in future commits.

pohly replied:
Done.

dgrisonnet commented:
/triage accepted
/assign @logicalhan

logicalhan reviewed (approved):
/lgtm
/approve
(from sig-instrumentation)

alculquicondor commented:
/hold
for scheduling review

pohly commented:
I've pushed the fix: https://github.com/kubernetes/kubernetes/compare/6e608b46e4f69344f98d065007e05406547be764..7f8e72a2c9c1cc728c9fa1aa86d442915d658c5f

alculquicondor commented on test/integration/scheduler_perf/config/performance-config.yaml:
```diff
@@ -649,3 +633,185 @@
+# with deterministic naming. The deterministic names are required
+# for the references between objects.
+#
+# They are currently commented out because of the additional time
```
maybe we can add some kind of flag "experimental" or "wip" that prevents them from running by default?

pohly replied:
Comment removed. I also renamed the test cases a bit: "InlineResourceClaim" wasn't quite accurate.

> Should we consider moving these tests into the dra folder?

I prefer to keep them in the main `performance-config.yaml` because then they'll automatically be included in the normal Prow jobs.

alculquicondor commented on test/integration/scheduler_perf/config/performance-config.yaml:
```diff
@@ -649,3 +633,185 @@
+      measurePods: 1000
+      maxClaimsPerNode: 10
+
+- name: LatencyWithInlineResourceClaims
```
why is this called "Latency"? Add a comment to explain the difference with the previous test.

pohly replied:
Added:

> This test case schedules one pod after the other and thus is dominated by the
> time that it takes to complete the scheduling of a single pod. In the other
> benchmarks, parallelism can hide the effect of delays per pod.

alculquicondor commented on test/integration/scheduler_perf/config/node-with-dra-test-driver.yaml:
```diff
@@ -0,0 +1,16 @@
+metadata:
+  generateName: scheduler-perf-dra-
+  labels:
+    have-test-resources: "true"
```
what's this label for? something that the test driver needs?

alculquicondor commented on test/integration/scheduler_perf/config/pod-with-claim-reference.yaml:
```diff
@@ -0,0 +1,15 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  Name: test-dra---counter
```
why counter? Also, shouldn't this use generateName? But probably not `Name` (uppercase N).

pohly replied:
I needed a special string that is accepted inside YAML.

I chose `counter` because its a running counter starting at zero inside the current operation. I chose `--counter` because `<counter>` was harder to use in JSON (gets encoded differently).

Using `generateName` defeats the purpose of having deterministic names. But why does `Name` work? Interesting. It looks like it doesn't matter. The Go field is properly populated.

pohly replied:
I'm replacing `Name` with `name`, of course - it's still the right spelling for YAML.

alculquicondor commented on test/integration/scheduler_perf/gnuplot.sh:
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env bash
+
+# Copyright 2016 The Kubernetes Authors.
```
2023 :)

pohly replied:
Time flies... fixed.

alculquicondor commented on test/integration/scheduler_perf/gnuplot.sh:
```diff
@@ -0,0 +1,47 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Invoke this script with a list of *.dat and it'll plot them with gnuplot.
```
Add a note about how to use this script in combination with go test in the readme.

pohly replied:
Done.

alculquicondor commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -223,6 +235,10 @@ func (op *op) UnmarshalJSON(b []byte) error {
 		&createNamespacesOp{},
 		&createPodsOp{},
 		&createPodSetsOp{},
+		&createResourceClaimsOp{},
+		&createResourceClaimTemplateOp{},
+		&createResourceClassOp{},
+		&createResourceDriverOp{},
```
Any chance these can all be replaced with a single createObjectsFromTemplate?

pohly replied:
I could use Go generics?

alculquicondor commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -384,15 +407,18 @@ type createPodsOp struct {
+	// deleted one after the other. This can be used to determine
+	// the latency for getting a single pod running instead of
+	// pod scheduling throughput.
+	Sequentially bool
```
Can you use the latency metrics instead? In particular `pod_scheduling_duration_seconds`

pohly replied:
Yes, `pod_scheduling_duration_seconds` seems better than observing the same thing in the client (`PodLatency`). That will make the code quite a bit simpler. Just to be sure before I do that, `pod_scheduling_duration_seconds` tracks the entire time for a pod also across multiple scheduling attempts, right?

The numbers match, so it looks like it does.

pohly replied:
I removed `PodLatency`. But `createPodsSequentially` remains because creating pods in bulk vs. one-after-the-other are different kind of workloads.

pohly replied:
But doing it via qps then is just rate limiting. It doesn't create the next pod as soon as the old one is scheduled.

That answers the question "can I schedule x pods/s
?". It does not answer the question "on average, how long does it take to schedule one pod?". For that, clusterloader would have to increase the rate until the cluster doesn't keep up anymore.

alculquicondor replied:
> on average, how long does it take to schedule one pod?

You shouldn't need to create pods sequentially to measure this. That's why we have latency metrics. Creating pods sequentially sounds very artificial (detached from reality in production).

A more appropriate question is how long do pods take to schedule on average at a given pod creation rate.

pohly replied:
We seem to be interested in different answers. When I was working on reducing latency, it was easier to deal with scheduling of one pod at a time. To achieve that with a fixed pod creation rate, that rate would have to be low enough to not accidentally create the next pod too soon, so most of the time it'll run more slowly than a test that creates the next pod as soon as possible.

I can keep this in a private branch if no-one else is interested in this.

alculquicondor replied:
I just think it's too artificial to be worth

alculquicondor replied:
users in a real cluster don't wait for other pods to be scheduled before creating their own.

alculquicondor commented on test/integration/scheduler_perf/util.go:
```diff
@@ -290,6 +341,10 @@ type throughputCollector struct {
+	startScheduled, endScheduled int
 }
 
 func newThroughputCollector(podInformer coreinformers.PodInformer, labels map[string]string, namespaces []string) *throughputCollector {
```
-1 to this change. I think you are mixing latency and throughput in a single function.

pohly replied:
"this change" = a1a92e5601cb11e9b31f3c0cab6cf1bd238f0c1b = `scheduler_perf: also record metric for "total pods/s"`?

I wouldn't say that this adds latency measurements. It's still the same metric (pods/s), just determined over the entire duration of the measurement.

In practice, it's close to the Perc99:
```
name                                                           SchedulingThroughput/Average
PerfScheduling/SchedulingBasic/500Nodes-36                                                                  1.00k ± 0%
PerfScheduling/SchedulingPodAntiAffinity/500Nodes-36                                                          200 ± 0%
PerfScheduling/SchedulingSecrets/500Nodes-36                                                                  724 ± 0%
PerfScheduling/SchedulingInTreePVs/500Nodes-36                                                                364 ± 0%
PerfScheduling/SchedulingMigratedInTreePVs/500Nodes-36                                                        363 ± 0%
PerfScheduling/SchedulingCSIPVs/500Nodes-36                                                                   350 ± 0%
PerfScheduling/SchedulingPodAffinity/500Nodes-36                                                              250 ± 0%
PerfScheduling/SchedulingPreferredPodAffinity/500Nodes-36                                                     500 ± 0%
PerfScheduling/SchedulingPreferredPodAntiAffinity/500Nodes-36                                                 500 ± 0%
PerfScheduling/SchedulingNodeAffinity/500Nodes-36                                                             537 ± 0%
PerfScheduling/TopologySpreading/500Nodes-36                                                                  500 ± 0%
PerfScheduling/PreferredTopologySpreading/500Nodes-36                                                         500 ± 0%
PerfScheduling/MixedSchedulingBasePod/500Nodes-36                                                             333 ± 0%
PerfScheduling/PreemptionPVs/500Nodes-36                                                                     45.4 ± 0%
PerfScheduling/Unschedulable/500Nodes/200InitPods-36                                                          500 ± 0%

...

name                                                           SchedulingThroughput/Perc99
PerfScheduling/SchedulingBasic/500Nodes-36                                                                  1.00k ± 0%
PerfScheduling/SchedulingPodAntiAffinity/500Nodes-36                                                          295 ± 0%
PerfScheduling/SchedulingSecrets/500Nodes-36                                                                  908 ± 0%
PerfScheduling/SchedulingInTreePVs/500Nodes-36                                                                511 ± 0%
PerfScheduling/SchedulingMigratedInTreePVs/500Nodes-36                                                        485 ± 0%
PerfScheduling/SchedulingCSIPVs/500Nodes-36                                                                   567 ± 0%
PerfScheduling/SchedulingPodAffinity/500Nodes-36                                                              346 ± 0%
PerfScheduling/SchedulingPreferredPodAffinity/500Nodes-36                                                     606 ± 0%
PerfScheduling/SchedulingPreferredPodAntiAffinity/500Nodes-36                                                 586 ± 0%
PerfScheduling/SchedulingNodeAffinity/500Nodes-36                                                             985 ± 0%
PerfScheduling/TopologySpreading/500Nodes-36                                                                  549 ± 0%
PerfScheduling/PreferredTopologySpreading/500Nodes-36                                                         616 ± 0%
PerfScheduling/MixedSchedulingBasePod/500Nodes-36                                                             532 ± 0%
PerfScheduling/PreemptionPVs/500Nodes-36                                                                      429 ± 0%
PerfScheduling/Unschedulable/500Nodes/200InitPods-36                                                          528 ± 0%
```

Note that getting exactly 1000 pod/s for PerfScheduling/SchedulingBasic/500Nodes seems to be an artifact: I haven't checked, but I bet there's exactly one poll interval in which all pods get scheduled.

One can argue both ways: it's both redundant and provides a different view... I can take out the commit if it's too controversial.

alculquicondor replied:
> One can argue both ways: it's both redundant and provides a different view... I can take out the commit if it's too controversial.

Not sure if it is. This PR is so long that it's hard to tell what is required and what is a new test feature. If you can remove it for now, I'd appreciate.

I know you tried to split everything into different commits, but as long as github doesn't show incremental updates well enough, it doesn't help much.

pohly replied:
I'll try to strip out as many changes as possible from this PR.

kerthcet commented:
I'll also take a look.

alculquicondor commented on test/integration/framework/perf_utils.go:
```diff
@@ -89,7 +91,20 @@ func (p *IntegrationTestNodePreparer) PrepareNodes(ctx context.Context, nextNode
+			node := baseNode.DeepCopy()
+			name := node.Name
+			if name == "" {
+				name = node.GenerateName + fmt.Sprintf("%x", rand.Int31())
```
use `String` in `apimachinery/pkg/util/rand`

pohly replied:
Done.

pohly commented on test/integration/scheduler_perf/util.go:
```diff
@@ -75,17 +81,26 @@ func newDefaultComponentConfig() (*config.KubeSchedulerConfiguration, error) {
 // Notes on rate limiter:
 //   - client rate limit is set to 5000.
-func mustSetupScheduler(ctx context.Context, b *testing.B, config *config.KubeSchedulerConfiguration) (util.ShutdownFunc, coreinformers.PodInformer, clientset.Interface, dynamic.Interface) {
+func mustSetupScheduler(ctx context.Context, b *testing.B, config *config.KubeSchedulerConfiguration, withDynamicResourceAllocation bool) (informers.SharedInformerFactory, clientset.Interface, dynamic.Interface) {
```
Map of features - that's what the caller has.

pohly commented on test/integration/scheduler_perf/util.go:
```diff
@@ -107,15 +122,30 @@ func mustSetupScheduler(ctx context.Context, b *testing.B, config *config.KubeSc
-		tearDownFn()
+	// Testing of DRA with inline resource claims depends on these
+	// controllers for creating and removing ResourceClaims.
+	if withDynamicResourceAllocation {
```
I put this here because of `util.StartFakePVController`. That is another additional controller run in the background by `mustSetupScheduler`.

A separate helper function for the code makes sense. I don't care where that then gets called, but it should be consistent with calling `util.StartFakePVController`. Do you prefer both in `mustSetupScheduler` or both in `runWorkload`?

pohly commented on test/integration/scheduler_perf/config/pod-with-resource-claim-template.yaml:
```diff
@@ -0,0 +1,15 @@
+apiVersion: v1
```
Done.

pohly commented:
I've pushed an update which addresses the review comments by updating the commits: https://github.com/kubernetes/kubernetes/compare/7f8e72a2c9c1cc728c9fa1aa86d442915d658c5f..ff919e3ebd67e6eaa43a27ba79da0c44d3415a75

However, I also need to rebase because of a merge conflict.

pohly commented:
I've finished the rebase and double-checked that the DRA test cases still work.

kerthcet commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -937,7 +937,7 @@ func unrollWorkloadTemplate(b *testing.B, wt []op, w *workload) []op {
 }
 
-func runWorkload(ctx context.Context, b *testing.B, tc *testCase, w *workload) []DataItem {
+func runWorkload(ctx context.Context, b *testing.B, tc *testCase, w *workload, cleanup bool) []DataItem {
```
Should we make it an option? Considering we'll setup an etcd instance for each case, I think we can remove them directly. I created a separate PR here, but includes in this one also sgtm. You're the boss. https://github.com/kubernetes/kubernetes/pull/116843

pohly replied:
As I am not going to get this PR merged soon, I continued with that unit test in the same branch and just pushed an update. Perhaps review commit-by-commit and then before we merge, I take out everything after the commit which adds the DRA test cases and move the rest into a separate PR?

kerthcet commented:
/retest

@pohly are you still working on this?

pohly commented:
I don't have further changes planned for the commits at this point.

pohly commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -26,12 +26,17 @@ import (
 	"time"
 
+	"github.com/google/go-cmp/cmp"
+	"github.com/onsi/gomega"
```
Why?

Gomega is just another helper library for writing assertions. It works fine in Go unit tests.

The specific reason for using it here is that `gomega.Eventually` is vastly superior to `wait.Poll` ("timed out waiting for condition"...).

pohly commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -699,6 +904,93 @@ func BenchmarkPerfScheduling(b *testing.B) {
 	}
 }
 
+var testSchedulingLabelFilter = flag.String("test-scheduling-label-filter", "unittest", "comma-separated list of labels which a testcase must have (no prefix or +) or must not have (-)")
```
> I don't think these are unit tests. 

Let's replace "unittest" with "functionality" or "integration-tests".

> But we have dedicated integration tests. Why add this?

Two reasons:
- It ensures that the code going into scheduler_perf gets tested in a pre-merge Prow job. Previously, linting only ensured that it compiles, but nothing actually ran the code.
- Last but not least, it reduces the amount of test code that needs to be written for new features like DRA. The same test can cover correctness and performance checks, depending on the configuration. Simulating special scenarios may still need separate tests, but the basic "does a pod get scheduled with dynamic resource claim" is covered well here already.

I was about to add that race detection can be enabled when performance isn't relevant (there were bugs in scheduler_perf found by that), but "pull-kubernetes-integration" currently doesn't do that. I'm checking in #sig-testing whether it could get enabled.

pohly commented:
I still have all commits in this one PR, including the ones that don't belong here.

Once the branch is open, I will create separate PRs for some of the preparation commits and we can merge those in smaller chunks.

alculquicondor commented:
Thanks!

pohly commented:
Finally! :grin: 

There's now only commit left in this PR, the one which adds support for DRA.

@alculquicondor, @kerthcet: please review.

pohly commented on test/integration/util/util.go:
```diff
@@ -102,10 +103,22 @@ func StartScheduler(ctx context.Context, clientSet clientset.Interface, kubeConf
 // TODO(mborsz): Use a real PV controller here.
-func StartFakePVController(ctx context.Context, clientSet clientset.Interface) {
-	informerFactory := informers.NewSharedInformerFactory(clientSet, 0)
+func StartFakePVController(ctx context.Context, clientSet clientset.Interface, informerFactory informers.SharedInformerFactory) {
```
This probably should have been in https://github.com/kubernetes/kubernetes/pull/117232...

pohly commented:
/hold cancel
